Core/src/IfcGeometryConverter/FaceConverter.h

```diff
@@ -801,59 +790,73 @@ namespace OpenInfraPlatform {
-					std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>> ifcFaceBounds,
-					const carve::math::Matrix& pos,
-					std::vector<std::vector<carve::geom::vector<3>>>& faceBoundLoops) const noexcept(true)
+				std::vector<std::vector<carve::geom::vector<3>>> convertIfcFaceBoundList(
+					const std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>>& ifcFaceBounds,
+					const carve::math::Matrix& pos) const noexcept(true)
+				{
+					const std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>> faceBounds = swapOuterBoundaryToFront(ifcFaceBounds);
+
+					std::vector<std::vector<carve::geom::vector<3>>> faceBoundLoops;
+					faceBoundLoops.resize(faceBounds.size());
+					for (size_t i = 0; i < faceBounds.size(); i++)
+					{
+						// get vertices of one bound loop into faceBoundLoops[i]
+						faceBoundLoops[i] = convertIfcFaceBound(faceBounds[i], pos);
+					}
+
+					return faceBoundLoops;
+				}
+
+				std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>> swapOuterBoundaryToFront(
+					std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>> ifcFaceBounds) const throw(...)
 				{
 					// As carve expects outer boundary of face to be at first index, outer boundary has to be moved to index 0 and inner boundary has index >= 1
 					for (auto it = ifcFaceBounds.begin(); it != ifcFaceBounds.end(); it++)
 					{
 						if (it->isOfType<typename IfcEntityTypesT::IfcFaceOuterBound>())
 						{
-							// swap element 'it' with first element
+							// swap element 'it' with first element;
 							std::rotate(ifcFaceBounds.begin(), it, it + 1);
 
 							// according to ifc-documentation, only one boundary (= loop) can be outer boundary:
 							// "One loop is optionally distinguished as the outer loop of the face."
 							// (https://standards.buildingsmart.org/IFC/DEV/IFC4_3/RC3/HTML/link/ifcface.htm)
 							// thus, break the for-loop after THE outer boundary was found
-							break;
+							return ifcFaceBounds;
 						}
 					}
-
-					faceBoundLoops.resize(ifcFaceBounds.size());
-					for (size_t i = 0; i < ifcFaceBounds.size(); i++)
-					{
-						// get vertices of one bound loop into faceBoundLoops[i]
-						convertIfcFaceBound(ifcFaceBounds[i], pos, faceBoundLoops[i]);
-					}
 				}
 
-				void convertIfcFaceBound(
+				std::vector<carve::geom::vector<3>> convertIfcFaceBound(
 					const EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>& bound,
-					const carve::math::Matrix& pos,
-					std::vector<carve::geom::vector<3>>& loopVertices3D) const noexcept(false)
+					const carve::math::Matrix& pos) const noexcept(false)
 				{
+					if (bound.expired()) {
+						throw oip::ReferenceExpiredException(bound);
+					}
+
 					//	IfcLoop   has subtypes   IfcEdgeLoop, IfcPolyLoop, IfcVertexLoop
 					const EXPRESSReference<typename IfcEntityTypesT::IfcLoop>& loop = bound->Bound;
 					bool polyOrientation = bound->Orientation;

```

Comment:
do we react to this boolean?

Reply:
A few lines later, it's used to flip the orientation / sense of the boundary.
https://github.com/tumcms/Open-Infra-Platform/blob/2715f2c87179552ecdd25ba656edf031af0fa9cb/Core/src/IfcGeometryConverter/FaceConverter.h#L849-L853
But the variable is only used once, thus commit 3aab188 would replace the variable with its actual information source.